### PR TITLE
search: Avoid excessive rerendering of search sidebar sections

### LINKED ISSUE #24996: search: Avoid completely rerendering the sidebar on every keypress
We are currently rerendering the sidebar on every keypress in the search query input (this is working off #24844 to demonstrate that the sidebar doesn't rerender because the whole page rerenders): 

https://user-images.githubusercontent.com/179026/133458212-2210c6f8-308c-4cbc-b9f8-cc36045bcd78.mp4

This is mainly due to how we render the various links/buttons: They all either directly use the current query or depend on an event handler that uses the current query (and thus gets updated every time the query changes).

A simple solution would be to use `useRef` with the query so that the event handler doesn't have to be recreated. But I would like to explore one of these solutions:

- Add a mini redux-like actions system to mutate the query. That means instead of manipulating the query directly, create mutation "actions" (update filter, add filter, toggle filter, etc) that is passed to the query store. This is certainly a more drastic approach but decouples everything from the current query.
- Use an immutable "`updateQuery`", which accepts a callback which gets passed the current query, in/as handlers so that they don't have to be recreated every time the query changes. Kind of like `useState` accepts a callback to pass in the current state.

### PR DESCRIPTION
**Pull request by fkling**
Resolves #24996

tl;dr: Sidebar renders now 20x faster on every keypress

(note: enable "hide whitespace changes")

This PR does a couple of things:

- Adds a new method (action) to `NavbarQueryState` which makes it possible to submit a new search query without needing a reference to the current query. This means that components don't need to receive the current query as props, which means they don't need to rerender on every query update.
- A couple of things have been memoized, especially `SearchReference` because it contributes the most to sidebar render time (maybe there is room for improvement there as well). The general advice is to apply memoization carefully. I think it in this case it is necessary because the search results page is one of the most important pages and typing into the search input is the main action one does on that page. Some of the memoization might not be necessary if we move more query related state out of the root component state (such as `patternType`).
- Cleaned up props so that really only what is needed is passed to the search sidebar (also I added a bunch of unnecessary things to `SearchReference` because I didn't know what I was doing.

Especially the changes to `NavbarQueryState` shows how we can consolidate query related logic into one place. There are more opportunities to improve this.

Before: 

https://user-images.githubusercontent.com/179026/133619581-d4a7d766-801f-4100-b468-fce09b05c909.mp4

After:

https://user-images.githubusercontent.com/179026/133619625-b0055ce9-11b0-4ff2-a956-b311b48f8692.mp4

### CODE REVIEW

File: client/web/src/search/results/StreamingSearchResults.tsx [hunk 1]
```diff
@@ -237,7 +240,13 @@ export const StreamingSearchResults: React.FunctionComponent<StreamingSearchResu
             <PageTitle key="page-title" title={query} />
 
             <SearchSidebar
-                {...props}
```
Comment (limitedmage) [inline client/web/src/search/results/StreamingSearchResults.tsx:240]: Does this really make a difference in performance? Should we consider making `{...props}` an antipattern and prohibit it with a linter rule?

File: client/web/src/search/results/sidebar/SearchSidebar.tsx [hunk 2]
```diff
@@ -169,17 +178,19 @@ export const SearchSidebar: React.FunctionComponent<SearchSidebarProps> = props
                 ) : null}
                 {repoName ? (
                     <SearchSidebarSection
+                        id={SectionID.REVISIONS}
```
Comment (limitedmage) [inline client/web/src/search/results/sidebar/SearchSidebar.tsx:181]: I'm not sure if I like the IDs using very generic words like "types" or "repositories", it seems like it can cause collisions. Ideally, we should avoid IDs in HTML elements unless absolutely necessary (eg. accessibility requirements). Can this be solved without using IDs? If IDs are required, I would pre-pend something to the IDs (eg. `search-reference-panel-${SectionID.REPOSITORIES`)

(If this is not meant to be an HTML ID, ignore the above and I would instead rename the prop to something like `sectionId` instead to avoid confusion with HTML ID property.)

**Review by limitedmage — APPROVED** [review]
Looks good overall, left some comments and suggestions.

File: client/web/src/search/results/StreamingSearchResults.tsx (hunk 1 above)
Comment (fkling) [reply client/web/src/search/results/StreamingSearchResults.tsx:240]: It probably doesn't make a difference in this specific case, but personally I do think that we should treat `{...props}` as an antipattern. Maybe something to bring up in the next frontend crew meeting?

File: client/web/src/search/results/sidebar/SearchSidebar.tsx (hunk 2 above)
Comment (fkling) [reply client/web/src/search/results/sidebar/SearchSidebar.tsx:181]: Ah, thank you for pointing this out. Yes, these are not HTML IDs. I'll make that clearer!
